### DISCUSSION TITLE
unable to get quicksync working in Ubuntu 20.04 + docker

### DISCUSSION THREAD

**Discussion by mal5305** [blakeblackshear/frigate#3777] [Ask A Question]
Hardware: Celeron G5905 (Comet Lake/10th Gen), Coral USB, combination of Unifi, Reolink, Amcrest cameras
Software: Ubuntu 20.04, Frigate latest as of 2022-09-06

I'm unable to get any camera to run with Quicksync after trying numerous configurations.
```
ls /dev/dri
by-path  card0  renderD128
```
Docker-compose snip for devices:
```
devices:
      - /dev/bus/usb:/dev/bus/usb # passes the USB Coral, needs to be modified for other versions
      - /dev/dri/renderD128 # for intel hwaccel, needs to be updated for your hardware
```
Amcrest camera config:
```
  backyard:
    ffmpeg:
      hwaccel_args:
      - -hwaccel
      - qsv
      - -hwaccel_device
      - /dev/dri/renderD128
      inputs:
        - path: rtsp://user:password@192.168.20.63:554/cam/realmonitor?channel=1&subtype=0
          roles:
            - rtmp
```
I don't see any errors in the docker logs, but the output of `intel-gpu-top` shows no quicksync activity, and the ffmpeg process is using quite a bit of CPU usage.

It's very possible I'm totally misunderstanding some core concept, but I'm kind of stuck here. I'm happy to provide any additional information, I'm just not sure what to try next. Thank you!

**Comment by NickM-27** [comment]
What version of frigate are you on? Have you tried to see if VAAPI works?

**Reply by mal5305** [reply]
> What version of frigate are you on? Have you tried to see if VAAPI works?

Latest: (0.10.1)

I've tried VAAPI according to the docs but it failed with this message: `[h264 @ 0x556eed144f80] No device available for decoder: device type vaapi needed for codec h264.`

I used these config items:
```
ffmpeg:
  hwaccel_args:
    - -hwaccel
    - vaapi
    - -hwaccel_device
    - /dev/dri/renderD128
    - -hwaccel_output_format
    - yuv420p
```

**Reply by NickM-27** [reply]
Interesting, in that case I'd recommend using frigate 0.11 RC. It is using a new ffmpeg build and version which may work better. Please be sure to read the release notes and breaking changes.

**Comment by moandcompany** [comment]
Have you tried the following `hwaccel_args` for your camera stream? 

The following configuration works for Frigate 0.10.1 (stable) on Ubuntu 10.04 with Intel QuickSync (8th gen Coffee Lake CPU/iGPU)

```
ffmpeg:
  hwaccel_args:
     - -hwaccel vaapi
     - -hwaccel_device /dev/dri/renderD128
     - -hwaccel_output_format yuv420p
```

```
cameras:
  camera_name:
    ffmpeg: 
      inputs:
        - path:  {CAMERASTREAM_URI_HERE}
          roles:
            - detect
      hwaccel_args:
        - -c:v
        - h264_qsv
```

Also check to make sure the VAAPI drivers are installed for your Intel iGPU:

- `apt info intel-media-va-driver` or `apt info intel-media-va-driver-non-free`

You can also use `vainfo` to see and verify that your drivers are installed. Example output below:

```
libva info: VA-API version 1.14.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_14
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.14 (libva 2.14.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 22.4.3 (74f40ee)
vainfo: Supported profile and entrypoints
      VAProfileNone                   :	VAEntrypointVideoProc
      VAProfileNone                   :	VAEntrypointStats
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Simple            :	VAEntrypointEncSlice
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointFEI
      VAProfileH264Main               :	VAEntrypointEncSliceLP
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointFEI
      VAProfileH264High               :	VAEntrypointEncSliceLP
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointEncPicture
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264ConstrainedBaseline:	VAEntrypointFEI
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSliceLP
      VAProfileVP8Version0_3          :	VAEntrypointVLD
      VAProfileVP8Version0_3          :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointFEI
      VAProfileHEVCMain10             :	VAEntrypointVLD
      VAProfileHEVCMain10             :	VAEntrypointEncSlice
      VAProfileVP9Profile0            :	VAEntrypointVLD
      VAProfileVP9Profile2            :	VAEntrypointVLD

```

**Comment by mal5305** [comment]
ok, on the 0.11rc branch, paired with the config from @moandcompany (using vaapi) i believe it's working? `intel-gpu-top` shows activity in the Video engine:

![image](https://user-images.githubusercontent.com/3589639/188782593-27dbb27b-33f5-4719-9e8e-23414c420137.png)

however i'm still seeing high cpu usage for the ffmpeg process (~30%). i was under the impression that offloading to qsv would mean that ffmpeg would be very light on the CPU? am i misunderstanding a concept here?

**Reply by NickM-27** [reply]
It depends on a lot of things. Do you have a coral? How much motion do you have? What resolution are the streams you're dealing with?

Have you tried `hwaccel_args: -c:v h264_qsv` ?

**Reply by mal5305** [reply]
i have a coral, yes. but i haven't set up any motion detection or recording yet. this is more or less out of the box without any customizations besides hardware transcoding.

as far as resolution, i'm testing just one camera right now which is 1080p @ 15fps, around 4000 kbps

**Reply by NickM-27** [reply]
motion detection and object detection are enabled by default, so "I haven't set it up" doesn't mean it isn't currently on. 

I would try `hwaccel_args: -c:v h264_qsv` and see if the CPU usage is less.

**Reply by mal5305** [reply]
oh, well then i had no idea about the detection always being on, d'oh!

with the `h264_qsv` parameter, CPU usage sits around 20%

![image](https://user-images.githubusercontent.com/3589639/188983140-5fbdb3e1-09f8-4376-b247-d30c33ae643d.png)

The other thing I noticed is that there are 2 ffmpeg processes when I have one camera enabled. Maybe that's normal?

**Reply by NickM-27** [reply]
> oh, well then i had no idea about the detection always being on, d'oh!

to be clear, it can be turned off, it is just on by default

> with the h264_qsv parameter, CPU usage sits around 20%

seems like that is better than VAAPI then

> The other thing I noticed is that there are 2 ffmpeg processes when I have one camera enabled. Maybe that's normal?

The black text one you have highlighted is a process, meanwhile the green text one is a thread of that above process

**Reply by mal5305** [reply]
welp, i'm a bit of an idiot here. at least partially, and i'm not sure for how long... in these latest tests, i still had the `vaapi` parameters in the global ffmpeg config. i've been changing the config so much i'm not sure how long those globals have been set. so i've reverted to 
```
  hwaccel_args:
    - -hwaccel
    - qsv
    - -qsv_device
    - /dev/dri/renderD128
```
in global and i added `c:v` and `h264_qsv` to the camera config. the ffmpeg process is currently hovering in the 7-9% range with one camera set up. adding a second camera adds another process that takes up an additional 15-20% so it seems i have some tuning to do. i'm also running this on a dual core Celeron, so i will probably swap in a higher core-count i3.

thanks so much for your patience and assistance @NickM-27 and @moandcompany

**Comment by jvrobert** [comment]
Well, I had tried acceleration about a year ago (I have Alder Lake) and it was an epic fail. It actually did work for a bit, but then the GPU would crap out due to various bugs and I'd get green screens after between 30 minutes and a few hours.

Thinking everything would be fixed, I decided to try again - another epic fail. I upgraded Ubuntu to 22.04, tried latest 5.18 kernel, tried latest 5.19 kernel, used same (though probably newer) ffmpeg binaries on my host (for testing), installed latest GUC firmwares, added kernel options, installed various Intel media SDK libraries, etc.. Nothing. Errors of all sorts depending on what I did but nothing works.

Frigate == unmitigated awesomeness. Linux driver situation? Absolute epic embarrassment of a total failure. Just sad, really. Everyone involved should hang their heads in shame.

I will maybe try again in another year, fortunately my CPU is way overpowered so I don't really need the acceleration for my 4 cams..
